File: src/main/java/com/fengsheng/phase/ReceiveOrder.java
package com.fengsheng.phase;

import com.fengsheng.Player;
import com.fengsheng.card.Card;
import com.fengsheng.protos.Common;

import java.util.LinkedList;

/**
 * 角色接收第三张黑色情报的顺序
 */
public class ReceiveOrder extends LinkedList<Player> {
    public void addPlayerIfHasThreeBlack(Player player) {
        if (contains(player)) return;
        int count = 0;
        for (Card card : player.getMessageCards().values()) {
            for (Common.color color : card.getColors()) {
                if (color == Common.color.Black) count++;
            }
        }
        if (count >= 3) add(player);
    }

    public void removePlayerIfNotHaveThreeBlack(Player player) {
        int count = 0;
        for (Card card : player.getMessageCards().values()) {
            for (Common.color color : card.getColors()) {
                if (color == Common.color.Black) count++;
            }
        }
        if (count < 3) remove(player);
    }
}

File: src/main/kotlin/phase/ReceiveOrder.kt
package com.fengsheng.phase

import com.fengsheng.Player
import com.fengsheng.card.count
import com.fengsheng.protos.Common.color.Black
import java.util.*

/**
 * 角色接收第三张黑色情报的顺序
 */
class ReceiveOrder : LinkedList<Player>() {
    fun addPlayerIfHasThreeBlack(player: Player) {
        if (contains(player)) return
        val count = player.messageCards.count(Black)
        if (count >= 3) add(player)
    }

    fun removePlayerIfNotHaveThreeBlack(player: Player) {
        val count = player.messageCards.count(Black)
        if (count < 3) remove(player)
    }
}

interface HasReceiveOrder {
    val receiveOrder: ReceiveOrder
}